### ISSUE TITLE
TFMA not rendering in JupyterLab

### ISSUE THREAD

**Issue by ConverJens** [tensorflow/model-analysis#112]
### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow Model Analysis)**: Yes, minor
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 18.04 and image python:3.7-slim
-   **TensorFlow Model Analysis installed from (source or binary)**: pip install
-   **TensorFlow Model Analysis version (use command below)**: 0.27.0
-   **Python version**: 3.7
-   **Jupyter Notebook version**: jupyterlab 2.2.9
-   **Exact command to reproduce**: see sample notebook

### Describe the problem

tfma.view.render_slicing_metrics shows no output.

### Source code / logs

Slim docker image to reproduce the issue:

```
FROM python:3.7-slim

ENV DEBIAN_FRONTEND=noninteractive

# This is used because our k8s cluster can only access our internal pypi
#COPY pip.conf /etc/pip.conf

# # TFMA is installed in the notebook because pip complained otherwise
RUN python3.7 -m pip install --no-cache-dir jupyterlab==2.2.9

# Install Node (for jupyter lab extensions)
RUN apt update && \
    apt -y install nano curl dirmngr apt-transport-https lsb-release ca-certificates && \
    curl -L https://deb.nodesource.com/setup_15.x | bash - && \
    apt update && apt install -y nodejs && \
    node -v

RUN jupyter labextension install tensorflow_model_analysis@0.27.0 && \
    jupyter labextension install @jupyter-widgets/jupyterlab-manager@2

RUN jupyter lab build

ENV DEBIAN_FRONTEND=

ENV NB_PREFIX /
ENV SHELL=/bin/bash

# Standard KubeFlow jupyter startup command
CMD ["/bin/bash","-c", "jupyter lab --notebook-dir=/home/jovyan --ip=0.0.0.0 --no-browser --allow-root --port=8888 --NotebookApp.token='' --NotebookApp.password='' --NotebookApp.allow_origin='*' --NotebookApp.base_url=${NB_PREFIX}"]
```

Below is an evaluation artifact from a small TFX pipeline and a minimal notebook to reproduce. Notebook consists of unzipping, install tfma, load eval result and try to display. 
[3053.zip](https://github.com/tensorflow/model-analysis/files/6010404/3053.zip)
[tfma-render-issue.ipynb.zip](https://github.com/tensorflow/model-analysis/files/6010408/tfma-render-issue.ipynb.zip)

**Comment by fhuanming** [comment]
cc: @atn832

**Comment by atn832** [comment]
Thanks for sharing the steps. I haven't tried reproducing yet, but judging from the error you posted at https://github.com/tensorflow/model-analysis/issues/56#issuecomment-780372449, it seems like the TFMA extension is downloading vulcanized_tfma.js from the wrong URL. As a result, instead of actual Javascript, it gets some kind of markup, probably something like:

```html
<!doctype html>
<html>
  ...
  file not found
  ...
</html>
```

It reads the first `<` and returns this error:

> Uncaught SyntaxError: Unexpected token '<' vulcanized_tfma.js:1

In the Network tab of the Chrome debugger tool, can you check the url of the request for vulcanized_tfma.js and share it here?

**Comment by ConverJens** [comment]
@atn832  If I understand correctly, the url is: <kubeflow-url>:31380/nbextensions/tensorflow_model_analysis/vulcanized_tfma.js

Is TFMA trying to download anything? We are running in an on-prem cluster with no external access so if TFMA needs external access that would be an issue.

**Comment by ConverJens** [comment]
@atn832 @mdreves Here is the output from stats gen:
![Screenshot 2021-02-22 at 14 55 52](https://user-images.githubusercontent.com/61828156/108717947-2a0e9700-751e-11eb-9df9-be97004cf779.png)

**Comment by Bobgy** [comment]
Getting the same issue in https://github.com/kubeflow/pipelines/issues/5194, we use technique in https://github.com/tensorflow/model-analysis/issues/10#issuecomment-595618035 to visualize TFMA as html and then embed that HTML in a different place in iframe.

The problem is similar to this reported issue that vulcanized_tfma.js doesn't load. I verified that TFMA 0.26.0 still works for us, so the regression happens between the two versions.

**Comment by ConverJens** [comment]
@Bobgy Is the trick with embedding the html required? This is not something I usually use, for instance when rendering stats from TFDV.

**Comment by Bobgy** [comment]
My use case

I generate the html in a step in a pipeline, and the pipeline UI shows the html in an iframe. but that's not related to this issue

**Comment by fhuanming** [comment]
Hi @ConverJens, besides what @atn832 asked, I have run the dockerfile with the notebook and data that you shared above (by `docker run -p 8888:8888 {image_name}`).  I can successfully load the TFMA UI on the JupyterLab. 

In terms of
> Is TFMA trying to download anything?

I don't know what the answer is. Maybe @atn832 can give the answer. I only know that running `jupyter labextension install tensorflow_model_analysis@0.27.0` will download TFMA js packages from NPM.

**Comment by atn832** [comment]
Upon running `render_slicing_metrics` in a cell, Chrome will download vulcanized_tfma.js. This is hosted by Jupyter Lab once you install the TFMA extension, so it should work even without internet.

![image](https://user-images.githubusercontent.com/3399854/109597192-16d17c00-7b5b-11eb-8dd1-ca75a0541c87.png)

@ConverJens, can you share what you see in the Response tab for the vulcanized_tfma.js request? I expect it'll show some HTML with an error message that might tell us what is wrong.

On my machine, since it worked, I can see a bunch of Javascript comments followed by Javascript code like this:
![image](https://user-images.githubusercontent.com/3399854/109597260-3cf71c00-7b5b-11eb-81ac-5fa815e6026a.png)

**Comment by ConverJens** [comment]
@atn832 You are completely right! The response I get is basically `Please enable JavaScript to view this website.` which is very weird since my chrome settings says it's allowed for all sites, which is the recommended setting. I'm using latest version of Chrome (88). Full response below. I get the same response when using Safari.

Any idea how to proceed?

@Bobgy You mentioned that you had this issue in KubeFlow as well. Is there some interaction with in KF that leads the browser into believing that JavaScript is disabled or otherwise may be causing this issue?

```
<!doctype html><html lang="en"><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1,user-scalable=yes"><meta name="description" content="Kubeflow Central Dashboard"><meta name="theme-color" content="#3f51b5"><title>Kubeflow Central Dashboard</title><link rel="shortcut icon" href="/assets/favicon.ico"><link rel="icon" href="/assets/favicon-32x32.png" sizes="32x32"><link rel="icon" href="/assets/favicon-57x57.png" sizes="57x57"><link rel="icon" href="/assets/favicon-76x76.png" sizes="76x76"><link rel="icon" href="/assets/favicon-96x96.png" sizes="96x96"><link rel="icon" href="/assets/favicon-128x128.png" sizes="128x128"><link rel="icon" href="/assets/favicon-192x192.png" sizes="192x192"><link rel="apple-touch-icon" href="/assets/favicon-152x152.png" sizes="152x152"><link rel="apple-touch-icon" href="/assets/favicon-180x180.png" sizes="180x180"><base href="/"><script src="webcomponentsjs/webcomponents-loader.js"></script><script src="webcomponentsjs/custom-elements-es5-adapter.js"></script><link href="app.css" rel="stylesheet"></head><body><main-page></main-page><noscript>Please enable JavaScript to view this website.</noscript><script src="vendor.bundle.js" defer="defer"></script><script src="app.bundle.js" defer="defer"></script><script src="dashboard_lib.bundle.js" defer="defer"></script></body></html>
```

**Comment by ConverJens** [comment]
@fhuanming @Bobgy Note that I'm rendering TFMA works for me locally as well when running the image and data I supplied, but the exact same rendering fails in my KubeFlow hosted notebook. I'm don't know if this issue is because of something in KubeFlow or from the fact that our k8s cluster has no external access.

@atn832 Do you know if TFMA needs to be able to reach NPM at runtime?

@Bobgy Any ideas if KubeFlow itself is blocking something?

**Comment by rcrowe-google** [comment]
@mwakaba2 the fact that you're getting a 404 makes me wonder about where it's trying to load vulcanized_tfma.js from.  Could you hover over the filename in the network request and give us the whole URL?

**Comment by mwakaba2** [comment]
Yeah that's the path.
```
/nbextensions/tensorflow_model_analysis/vulcanized_tfma.js
```
@Rcrow-google I think the issue is that our hosted notebook solution is using [jupyter_server ](https://github.com/jupyter-server/jupyter_server) in the backend. Jupyter_server by default doesn't support the `nbextensions` path. 
So to enable that, we have to enable [`nbclassic` server extension](https://github.com/jupyterlab/nbclassic).
 
I still get a 404 after that, so I think I also need to enable the extensions via `nbextension install`.

**Comment by mwakaba2** [comment]
Ok I got it working in Jupyterlab by installing and enabling the nbextensions.
```
 $ jupyter nbextension enable --py widgetsnbextension
 $ jupyter nbextension enable --py tensorflow_model_analysis
```
You need that in order to access the /nbextensions path.

**Comment by rcrowe-google** [comment]
Glad to hear you solved it Mariko!  It seems so simple once you know the answer ... 😁

**Comment by kwlzn** [comment]
@rcrowe-google there's still an upstream problem here FYI.

in the Jupyter ecosystem, there are two primary types of frontend extensions ([this doc may be helpful](https://blog.jupyter.org/99-ways-to-extend-the-jupyter-ecosystem-11e5dab7c54)):

* nbextensions - Jupyter Classic Notebook extensions - which support the classic Notebook UI.
* labextensions - Jupyter Lab Extensions - which support the modern Jupyter Lab UI.

nbextensions are effectively deprecated in favor of labextensions, because all modern notebook runtimes (incl AI Platform Notebooks etc are fronted with Jupyter Lab). so in order to enable nbextensions (Jupyter Classic Notebook extensions) on top of Jupyter Server (which replaces the Jupyter Notebook server) to support serving of this Javascript, we had to install a compatibility layer that otherwise wouldn't need to exist (called [nbclassic](https://github.com/jupyterlab/nbclassic)).

so tl;dr: the problem with TFMA's JupyterLab support is that it's not a pure JupyterLab extension. It's a labextension that _depends_ on a nbextension to work. what we've employed here is a workaround vs a proper fix - and this is likely where the other folks like @ConverJens are running into issues making this work.

**Comment by ConverJens** [comment]
@mwakaba2 @rcrowe-google I tried the fix that you mentioned but there was no difference for me, still no output. Albeit, the error I had was different from the one @mwakaba2 experienced.

@kwlzn I tried installing nbclassic but with no progress. I used to  start my server by running `jupyter lab ...` but should I start it with another command for this to have effect?

**Comment by mwakaba2** [comment]
@ConverJens 
You don't need nbclassic because jupyterlab 2.2.9 by default relies on the jupyter/notebook as the backend.
If you can access `/tree` then, there's no need for nbclassic. 
did you try installing the nbextensions in the docker image right after installing the lab extension?
```
jupyter nbextension install --py widgetsnbextension
jupyter nbextension enable --py widgetsnbextension
jupyter nbextension install --py tensorflow_model_analysis
jupyter nbextension enable --py tensorflow_model_analysis
```
My tfma setup is probably different from yours.
I used these instructions to install the tfma npm package. 
https://github.com/tensorflow/model-analysis/issues/56#issuecomment-780722092

**Comment by ConverJens** [comment]
@mwakaba2 
That was my understanding but I wanted to verify.

These are the commands I'm currently using:
```
RUN jupyter contrib nbextension install && \
    jupyter labextension install tensorflow_model_analysis@0.28.0 && \
    jupyter labextension install @jupyter-widgets/jupyterlab-manager@2  && \
    jupyter nbextension install --py --sys-prefix widgetsnbextension && \
    jupyter nbextension enable --py --sys-prefix widgetsnbextension && \
    jupyter nbextension install --py --sys-prefix tensorflow_model_analysis && \
    jupyter nbextension enable --py --sys-prefix tensorflow_model_analysis
```
along with pip installing the corresponding TFX version first. I've tested 0.27.0 and 0.28.0 with exactly the same result. I still cannot load the vulcanized_tfma.js file.

@kwlzn @rcrowe-google 
I'm not certain the issue that you are mentioning is actually what's causing my problem. I can run my image, notebook and data  locally successfully but once I host it, it fails. I can run it locally without the `jupyter nbextension` steps.

**Comment by ConverJens** [comment]
@atn832 @fhuanming @mwakaba2 @kwlzn @rcrowe-google 
So I think I found whats causing this issue: url rewrite in KubeFlow is causing the vulcanized_tfma path to point to the wrong place.

If I check the sources tab in the developer console and right click the vulcanized_tfma.js,  and choose "open in new tab" I'm directed to: <kubeflow host>/nbextensions/tensorflow_model_analysis/vulcanized_tfma.js which is incorrect since the notebook lives in: `<kubeflow host>/notebook/admin/<notebook server name>`.  If I manually concatenate these urls I get the actual js file: `<kubeflow host>/notebook/admin/<notebook server name>/nbextensions/tensorflow_model_analysis/vulcanized_tfma.js`

So it seems that this is actually a path issue.

**Comment by atn832** [comment]
Thank you for the details! We were able to reproduce the loading issue in our own environment and are working on a fix.

**Comment by jhamet93** [comment]
@atn832 

Out of curiosity, how is vulcanized_tfma.js produced ? I was not able to decipher it grepping through the project. 

It's ideal to drop the requirement of _nbclassic_ if a user is using JupyterLab since a user should be able to use this plugin with just JupyterLab.

**Comment by ConverJens** [comment]
@atn832 Fantastic, thank you!

@jhamet93 TFMA does not have nbclassic as a requirement, it was just proposed as an optional fix if you're running jupyter notebook as your backend instead of jupyter server.

**Comment by jhamet93** [comment]
@ConverJens The latest JupyterLab major version (which uses Jupyter Server as the backend) is planning on dropping the _nbclassic_ extension. Currently, there exists a shim to help ease users who are transitioning but this will cease to exist in the future. Thus, it makes sense for this to work out of the box with just JupyterLab since this is an antipattern. The _nbclassic_ plugin is only needed for the JupyterLab extension to serve a static file which should be able to be hosted by a different mechanism such as a server extension.

**Comment by ConverJens** [comment]
@atn832 Any update on this?

**Comment by ConverJens** [comment]
@atn832 Any update? I'll keep pinging :)

**Comment by atn832** [comment]
Yes, my fix was merged a few days ago (https://github.com/tensorflow/model-analysis/commit/cc7d75c1bf588123795511572e1b4445d9a52191) and this issue should be resolved on the next release.

**Comment by ConverJens** [comment]
Awsome! So it will be part of the 0.31.0 (or if it is the 1.0.0) release then?

**Comment by rclough** [comment]
@ConverJens according to the release notes @atn832 added to, it's in 0.29

**Comment by atn832** [comment]
Ah yes, I misplaced this release note! It should be part of 0.31.0 instead.

**Comment by mwakaba2** [comment]
@atn832 I'm running Jupyterlab in the Kubeflow Notebook Server and I'm still getting the "Please enable Javscript..." message. 

Because `isJupyterlab` is set to True, it looks like it's setting the wrong templatePath for Kubeflow.

<img width="1046" alt="Screen Shot 2021-05-28 at 5 11 45 PM" src="https://user-images.githubusercontent.com/3497137/120049601-c9d53880-bfd7-11eb-8d90-a28de69f496f.png">

```javascript
 // Jupyter Lab  
else if (window['isJupyterLab']) {    
     templatePath = '/nbextensions/tensorflow_model_analysis/';  
}  
// Kubeflow  
else {    
    templatePath = __webpack_public_path__;  
}
```

https://github.com/tensorflow/model-analysis/commit/cc7d75c1bf588123795511572e1b4445d9a52191#diff-736e4346295db6d4a1296db9cff539db3a62d53bce1431ede323fe1a91566eabR31

**Comment by atn832** [comment]
@mwakaba2, Thanks for reporting the issue and checking the flag. Even though you all had explained the problem well, I misunderstood it and fixed the wrong issue. I was fixing @Bobgy's use case posted on Feb 25 (https://github.com/tensorflow/model-analysis/issues/10#issuecomment-595618035), which is to show the TFMA UI in a standalone HTML page outside of a notebook environment, that HTML page being exported from a notebook.

This time, I'll see if I can support your use case of running the TFMA UI in Jupyter Lab hosted by Kubeflow Notebook Server.

**Comment by ConverJens** [comment]
@atn832 Thanks for keeping at it! This is a crucial piece to our user experience so thank you!

**Comment by mwakaba2** [comment]
@atn832 thank you for the clarification, and I look forward to the Kubeflow support!

**Comment by mwakaba2** [comment]
@atn832 Do you have any updates or know when this fix will be released?

**Comment by mwakaba2** [comment]
Hi @atn832, Sorry to bother you again. do you have any updates or ETA for the Kubeflow support? I'd like to make this extension available soon to Twitter employees using Kubeflow.

**Comment by ConverJens** [comment]
Ping @atn832. ETA?

**Comment by ConverJens** [comment]
@atn832 Any update on this? Did a fix make it into the 0.32.0 release?

**Comment by atn832** [comment]
Unfortunately I haven't got a chance to work on that yet. Keep you posted!

**Comment by mwakaba2** [comment]
@atn832 Do you have any updates?

**Comment by ConverJens** [comment]
@atn832 Update?

**Comment by ConverJens** [comment]
@atn832 I haven't given up you! Any update? @mdreves? Anyone else on the team? This is still a blocker for all folks using KubeFlow hosted notebooks.

**Comment by kwlzn** [comment]
I think @mwakaba2 was planning to take this on as a contrib?

**Comment by mwakaba2** [comment]
Yup, folks at Twitter are not able to use this extension for Kubeflow Notebooks. 
I think my team may be able to contribute sometime in Q4, but if Google can make an update before that, that would be ideal.

**Comment by atn832** [comment]
I won't be able to work on it after all. If @Bobgy has time, they might be able to fix it, since they have experience with the issue and know how to reproduce it. If not @mdreves or @rcrowe-google will have to find someone else.

**Comment by ConverJens** [comment]
Ping @Bobgy, are you up for it?

This is issue is still preventing us from actually using TFMA.

**Comment by mdreves** [comment]
We are understaffed in this area, but @Bobgy is going to try to take a look.

**Comment by ConverJens** [comment]
@mdreves That would be fantastic! @Bobgy do you have any idea when you will be able to start working on it? I would be happy to test it!

**Comment by Bobgy** [comment]
I'm starting to reproduce, just verified using a local jupyter lab instance with instructions like https://github.com/tensorflow/model-analysis/issues/112#issue-812044230. I can see tfma visualization in the notebook.

**Comment by Bobgy** [comment]
I reproduced the issue.
However, I think "Please enable Javascript" message is not an actual problem. https://github.com/tensorflow/model-analysis/issues/112#issuecomment-850715648
The root cause is that when downloading this file from for example: "https://dev-5-14.endpoints.gongyuan-dev.cloud.goog/nbextensions/tensorflow_model_analysis/vulcanized_tfma.js".
The content it downloads is actually kubeflow central dashboard HTML content.
What we expect is a minified javascript file for vulcanized_tfma.js.

Therefore, this seems like a route configuration problem that did not send the request "https://dev-5-14.endpoints.gongyuan-dev.cloud.goog/nbextensions/tensorflow_model_analysis/vulcanized_tfma.js" to the jupyter server.

**Comment by Bobgy** [comment]
I'm getting closer, the correct download url should be "https://dev-5-14.endpoints.gongyuan-dev.cloud.goog/notebook/gongyuan/tfma4/nbextensions/tensorflow_model_analysis/vulcanized_tfma.js"

**Comment by Bobgy** [comment]
NB_PREFIX env var is "/notebook/gongyuan/tfma4" for me case, so the correct URL is
HOST/NB_PREFIX/nbextensions/tensorflow_model_analysis/vulcanized_tfma.js.
Somehow tfma widget needs to know how to inject NB_PREFIX in there.

**Comment by Bobgy** [comment]
NB_PREFIX is passed to jupyterlab server in https://github.com/kubeflow/kubeflow/blob/f57be7ac6edc03561009eb3ae3f1344e16a05f09/components/example-notebook-servers/jupyter/s6/services.d/jupyterlab/run#L12

```
--ServerApp.base_url="${NB_PREFIX}"
```

**Comment by Bobgy** [comment]
That's pretty much what I can figure out based on my knowledge.
@mdreves can you or someone you know help figure out how `ServerApp.base_url` of jupyterlab can be passed to javascript in a jupyterlab widget? Then we need to fix https://github.com/tensorflow/model-analysis/blob/e3bb17a6f278191f87db8fc7b8831bbad5b9ed13/tensorflow_model_analysis/notebook/jupyter/js/lib/widget.js#L32-L34, so that it adds the base_url in front of the '/nbextensions/tensorflow_model_analysis/'.

**Comment by Bobgy** [comment]
Just inspecting on jupyterlab HTML, I found some config data

```
<script id="jupyter-config-data" type="application/json">{"appName": "JupyterLab", "appNamespace": "lab", "appSettingsDir": "/opt/conda/share/jupyter/lab/settings", "appUrl": "/lab", "appVersion": "3.0.12", "baseUrl": "/notebook/gongyuan/tfma4/" ... </script>
```

**Comment by ConverJens** [comment]
@mdreves I can probably test this during next week.

**Comment by ConverJens** [comment]
@mdreves 
I just tried to test this but when running `tfma.view.render_slicing_metrics(eval_result)` I got `Error displaying widget`. 
I'm using jupyterlab 3 and I have installed TFMA using the official instructions.
Any idea why I get this?

**Comment by ConverJens** [comment]
@mdreves  @jhamet93 @Bobgy @zywind Has anyone else tested this and got this to work in JupyterLab?

**Comment by zywind** [comment]
I've tried it and had the same error with 0.35 release.

**Comment by mdreves** [comment]
My apologies, but I don't have the cycles to work on this and we don't have anyone with expertise in this area at the moment. 

Just to clarify, it works except for when you go to display the results using `tfma.view.render_slicing_metrics`. Is the full error `Error displaying widget` or is there more information?

**Comment by zywind** [comment]
This is what I see in jupyter lab:
![Screen Shot 2021-11-17 at 3 43 26 PM](https://user-images.githubusercontent.com/3665540/142279328-f2f99f09-7455-4be8-9dee-2494883566ff.png)

**Comment by ConverJens** [comment]
@mdreves 
I see exactly the same as @zywind posted as well.

**Comment by rcrowe-google** [comment]
We'd be very interested if someone can do a PR to fix this.

**Comment by ConverJens** [comment]
@mdreves Do we have any update on this issue? Do you have the time to work on it now?

**Comment by mdreves** [comment]
@embr as FYI since he is the new TL for TFMA.

Unfortunately we still don't have the cycles to work on this and we don't have anyone with expertise in this area at the moment. We would be very happy to see a community contribution in this case.

**Comment by ConverJens** [comment]
@mdreves @embr 
Any update on this? Can this be prioritised now?

**Comment by ConverJens** [comment]
@JanetVictorious That's not fixing this at all, you've just just jupyter notebook instead of JupyterLab and this entire issue is about JupyterLab support.

**Comment by genehwung** [comment]
Hopefully someone has the expertise to work on a PR, meanwhile we are actively working on DataFrame API as an alternative to provide pandas DataFrame for easier plotting.

**Comment by rcrowe-google** [comment]
This may or may not help.  I recently got the TFMA visualization working in JupyterLab by changing my browser settings.  I'm using Chrome, and under Settings > Privacy and security > Security my setting was "Enhanced protection".  Changing it to "Standard protection" fixed the issue, which was caused by Chrome stopping what it sees as a CORB request.  I was using this code to display the visualization:

```
from IPython.display import IFrame
from ipywidgets.embed import embed_minimal_html

embed_minimal_html('tfma_slices_overview.html',
                   views=[store.display_tfma_analysis(13, slicing_column='trip_start_hour')],
                   title='Slices Overview')
```

```
IFrame(src='tfma_slices_overview.html', width=900, height=600)
```

I hope that helps!

**Comment by pindinagesh** [comment]
Hi @ConverJens 

Can you please confirm on the comment from @rcrowe-google  and suggest whether this can be closed. Thank you.

**Comment by janasangeetha** [comment]
Closing this due to inactivity. Feel free to reopen and post your comments(if you still have queries on this).
Thank you!
